src/cv_utils.cpp

```diff
@@ -57,6 +60,130 @@ void drawArrows(cv::Mat& image,
+
+    p1 = rect.tl();
+    cv::line(image, p1, p2, color);
+}
+
+void drawRotatedRect(cv::Mat& image,
+                     const cv::RotatedRect& rect,
+                     cv::Scalar color)
+{
+    cv::Point2f pts[4];
+    rect.points(pts);
+
+    cv::line(image, pts[0], pts[1], color, 3);
+    cv::line(image, pts[1], pts[2], color, 3);
+    cv::line(image, pts[2], pts[3], color, 3);
+    cv::line(image, pts[3], pts[0], color, 3);
+}
+
+bool insideImage(const cv::Size& image_size, int x, int y)
+{
+    return (x >= 0)
+        && (y >= 0)
+        && (x < image_size.width)
+        && (y < image_size.height);
+}
+
+bool insideRotatedRect(const cv::RotatedRect& rect, int x, int y)
+{
+    float rads = rect.angle * M_PI / 180;
+    cv::Vec2f dirx ( std::cos(rads), std::sin(rads));
+    cv::Vec2f diry (-std::sin(rads), std::cos(rads));
+
+    cv::Vec2f offset = cv::Point2f(x, y) - rect.center;
+    if (std::abs(dirx.dot(offset)) <= rect.size.width  / 2
+     && std::abs(diry.dot(offset)) <= rect.size.height / 2) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+void inRange(const cv::cuda::GpuMat& src,
+             cv::Scalar lowerb,
+             cv::Scalar upperb,
+             cv::cuda::GpuMat& dst,
+             InRangeBuf& buf)
+{
+    ROS_ASSERT(src.type() == CV_8UC3 || src.type() == CV_8UC4);
+
+    cv::cuda::split(src, buf.channels);
+
+    cv::cuda::subtract(cv::Scalar(255), buf.channels[0], buf.inverse);
+    cv::cuda::threshold(buf.inverse,
+                        buf.buf,
+                        255 - lowerb[0],
+                        255,
+                        cv::THRESH_BINARY_INV);
+    cv::cuda::bitwise_and(buf.buf, buf.buf, dst);
+    for (int i = 1; i < 3; i++) {
+        cv::cuda::subtract(cv::Scalar(255), buf.channels[i], buf.inverse);
+        cv::cuda::threshold(buf.inverse,
+                            buf.buf,
+                            255 - lowerb[i],
+                            255,
+                            cv::THRESH_BINARY_INV);
+        cv::cuda::bitwise_and(buf.buf, dst, dst);
+    }
+
+    for (int i = 0; i < 3; i++) {
+        cv::cuda::threshold(buf.channels[i],
+                            buf.buf,
+                            upperb[i],
+                            255,
+                            cv::THRESH_BINARY_INV);
+        cv::cuda::bitwise_and(buf.buf, dst, dst);
+    }
+}
+
+cv::Vec3d sumPatch(const cv::Mat& image, const cv::RotatedRect& rect)
+{
+    ROS_ASSERT(image.type() == CV_8UC3);
+
+    cv::Rect bounding_rect = rect.boundingRect2f();
+    size_t count = 0;
+    cv::Vec3d total (0, 0, 0);
+    for (int y = bounding_rect.y; y <= bounding_rect.y + bounding_rect.width; y++) {
+        for (int x = bounding_rect.x; x <= bounding_rect.x + bounding_rect.width; x++) {
+            if (cv_utils::insideImage(image.size(), x, y)
+             && cv_utils::insideRotatedRect(rect, x, y)) {
+                count++;
+                total += image.at<cv::Vec3b>(y, x);
```

Comment:
Everything in me screams vector instructions. I know that would be difficult because of the rotated rectangle though.

Reply:
Yeah same, I'm not doing that though.  I decided I'll make it efficient if it's a problem, seems like it isn't.